### DISCUSSION TITLE
Display translated fields in their own columns in the user's language

### DISCUSSION THREAD

**Discussion by DamienMrtl** [directus/directus#19761] [Feature Requests]
### Summary

Concerns #19626 and #19424 

So actually it is not possible to display a translated field in the user's language in its own column. It displays a comma separated list of all translated values. To display the translated value in the user's language, we have to use the "Translations" relational field and edit its display. But then, the column name is "Translations" so it does not correspond to the label of the field displayed. Also, if we have two translated fields, there is no way to display them in separate columns we are forced to joint them in the "Translations column".

### Basic Example

[2023-09-22-15-22-00.webm](https://github.com/directus/directus/assets/7550391/f3c5bee7-6fbe-4d59-b2d2-7b18bbc253fa)

### Motivation

It is complicated for the end user to use Directus when there are multiple translated fields.

### Detailed Design

To resolve this, I propose to create a new "Display" that we could enable on the fields in the translation relational collection. This display would allow displaying only the value in the current user's language, instead of the list of all translations.

But actually it is not possible to create a custom display doing this because custom displays does not have access to the other fields in the current item. So, no way to get the languages associated with a translated value. See #3583 

So I would like to implement some access to neighboring fields like state [here](https://github.com/directus/directus/discussions/3583#discussioncomment-7039525)

### Must Have:
- Display translated values in the current user's language in their own columns, so he can do sorting on the columns.
- Display the correct column label corresponding to the value in this column, so the user understands what is going on in the table view.

### Adoption Strategy

This would just add a new display, and new way to access fields values in custom displays, so no breaking changes here.

**Comment by DamienMrtl** [comment]
@rijkvanzanten So I made a PR to add neighboring field access for Displays #19876

**Reply by rijkvanzanten** [reply]
I appreciate the enthusiasm, but let's hold off on implementation / reviewing until we've gone through the feature request process: https://docs.directus.io/contributing/feature-request-process.html#request-a-feature. That ensures that we align the new features and engineering efforts with the bigger community requirements rather than individual needs. Thanks!

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by madc** [comment]
Has there been any development (or a workaround) on this issue? This would be very convenient..

**Reply by DamienMrtl** [reply]
Can you reopen my PR please ? @rijkvanzanten

**Comment by DamienMrtl** [comment]
My pull request was closed, maybe it should be reopened ? @rijkvanzanten

**Comment by T-Mike** [comment]
Hi,
I would like to second this and add some thoughts.

I understand, from a technical point of view, this is not a bug.
But, this is not the logical way the translations and labels should display in this context. The layout is very confusing for devs and clients and thus makes the whole translate feature unusable for real-world projects... which is frustrating because hey, we want to use directus for our projects! 🚀 

It seems from the outside that this problem has not been taken seriously because of the rapidly closed issues, the closed PR and the 'soon under review' status that has not been updated in 9 months although it gained some traction.
I imagine that you make the best out of the time and ressource you have but please take into consideration that this is not a just futile feature request.

**Comment by louisprp** [comment]
I agree completely, the translation feature is what really set directus apart for me compared to other alternatives, especially the two column editor. But then actually displaying the data in the UI feels clunky at best. I feel like there should be a way to have a configurable display field that has the ability to;

1. Display only the translation matching the current locale (with the option to have something like a tooltip to show the others - handy e.g. for titles) 
2. Display a computable amount of translations comma separated (this is how it’s shown currently if you select e.g. the Translatios->Title relation
3. Even better the option to switch the locale by clicking the button, or having an option where the sort by is located to quickly switch all entries between languages
4. 

Adding to this, I also noticed that it’s always a hassle to show the translations in text. Oftentimes it just says „2 Elements“ if I have two languages, instead of just showing the string values. This happens when I want to visualize an o2m relation in a table for example. 

I feel like the efforts made here are a good step in the right direction and find it unjustified to dismiss them. The goal is to enhance the user experience and providing the features that users need.

**Comment by vzkn** [comment]
Hi everyone,

First off, I want to say that our team absolutely enjoys working with Directus, it's truly a phenomenal product packed with so many useful features. Just yesterday, I tried out Directus v11, and I can't stress enough how much easier the new policies feature will make managing our users' permissions. It's a real game-changer! 🚀

I’d also like to share a few thoughts from my experience. I completely agree with the statements above and am thrilled that @DamienMrtl made a PR for this. It’s a bit of a bummer that it didn’t get merged earlier, but better late than never!

Recently, I discovered an interesting scenario when using this feature in an M2A relationship. We have a website builder tool based on Directus that utilizes content blocks, some of which include translations. For instance, our divider block has a translations field, and within it, there's a title field. We use the collection’s display template to show the title from translations (translations.title), as shown below:

![Example Image](https://github.com/user-attachments/assets/7d5f67fb-9370-4698-9763-3c8501e47be6)

Even though there are multiple translations:

![Multiple Translations](https://github.com/user-attachments/assets/3be1e1de-970a-4c8d-9de5-db180443e316)

This is how it looks for the users, everything appears clean and without unnecessary commas:

![User View](https://github.com/user-attachments/assets/1902c367-b079-4e40-a723-19ea10f772e3)

Our team would greatly appreciate getting this feature into production, as it would make it much more intuitive for content writers to navigate title and description fields, instead of translations field.

Thank you in advance!

**Reply by wottpal** [reply]
Couldn't upvote this enough. Isn't there any workaround right now? Relationship-fields targeting tables with translations are displayed horribly right now :/

![CleanShot 2024-10-23 at 11 45 49@2x](https://github.com/user-attachments/assets/225ec3d3-dcce-4c1d-b1e2-284bdd942b75)

![CleanShot 2024-10-23 at 11 45 28@2x](https://github.com/user-attachments/assets/4ff6c11e-a39c-4c3c-b1cd-26406757d2a8)

For us, it would be 100% fine to just display the value for a fixed language there, but right now it's not user friendly at all.

**Comment by louisprp** [comment]
@rijkvanzanten Any progress on this? As of v11, it seems that the same problem that is described in the original discussion still exists. I do not understand why the pull-request proposed cannot get merged to enable this feature, or if there is another way of doing it now. In any case, it is an issue to me that translations cannot be properly displayed in the currently selected language if you have more than one translated field and I feel that this really worsens the otherwise very great experience with translated fields.

It seems to me there are people willing to put in effort into achieving this, all that is needed is just the approval for the pull request?

**Comment by hsb-tonmoy** [comment]
This is clearly a widely wanted feature and it's weird how it's being ignored.

**Comment by azharul-islam** [comment]
@rijkvanzanten This feature would make translations much more useful. Now we see "2 items"(because we support two languages) in display and user has to go into every item to actually see the content. A preview would make a huge difference

**Comment by louisprp** [comment]
@DamienMrtl I was wondering if you’re still interested in developing for this. I’m currently trying out your solution and exploring an alternative way where the current row is also passed as a prop to the component that is rendered, but I believe there’s still work to be done. Also, did you achieve what you wanted in your original implementation? I couldn’t get it to also retrieve other fields alongside, but maybe I was doing it wrong. I would like to hear your thoughts on this!

**Comment by louisprp** [comment]
Just a quick update, I've achieved the display in the current language and this is what it currently looks like. There's still room for an improvement, as I would like to also add the current interface that is also shown with the regular translation component. Furthermore, for some reason it only works when the translations field is also in the row, not sure why that is the case, but at least this is a step in the right direction. This is based on the pull-request by @DamienMrtl, which works very well for this purpose, and I can imagine other use cases as well. @rijkvanzanten I believe this should provide a good basis for implementing this feature.


https://github.com/user-attachments/assets/ea51d52d-e4b7-4b19-89cc-ea770d3e0365

**Reply by louisprp** [reply]
For anyone looking to use my current implementation, you will have to first apply @DamienMrtl changes as in #19876, then you can add this as a component under `app/src/displays`

```
import { defineDisplay } from '@directus/extensions';
import { TYPES, LOCAL_TYPES } from '@directus/constants';
import { i18n } from '@/lang';

export default defineDisplay({
	id: 'mydisplay',
	name: 'Translated Values',
	icon: 'code',
	component: ({ value, field }) => {
		// Check if value is an array of translations
		if (Array.isArray(value)) {
			const activeLanguage = i18n.global.locale.value;
	
			// Find the translation that matches the active language
			const matchingTranslation = value.find(translation => translation.languages_code === activeLanguage);
	
			// Use the matching translation or fallback to the first translation
			const displayTranslation = matchingTranslation || value[0];
	
			// Return the name from the translation or a fallback message
			return displayTranslation[field] || null;
		}
			
		// If value is not an array, return it directly
		return value;
	},
	options: [],
	types: TYPES,
	localTypes: LOCAL_TYPES,
	localFields: ["languages_code"]
});
```

**Reply by wottpal** [reply]
Wow we need this to be merged asap 🔥

**Reply by QuentinSc** [reply]
WTF, it was never merged—how is that even possible?!

**Comment by louisprp** [comment]
Update on this: I've figured out how to get it working better now and have a nicer interface, it looks similar to the default translation preview now. It also works now when it's in another table through a relation. I've also fixed the issue that you needed to include other fields in the view, more on this below. The only thing that is left for me is to fix the view when displayed in a o2m table view (here it will still show 2 Elements since it seems to be a different rendering mechanism), but I guess this is a rather specific use-case. Here's a small video of what it looks like now and I'm pretty happy with the way it is right now:

https://github.com/user-attachments/assets/1bb95070-3fd0-4001-8778-fcc1e654e72f

In addition to the changes proposed in #19876, I've also added this code to `useAliasFields` in `use-alias-fields.ts`. This will ensure that the related collection is also included in the query, since there seemed to be issues with this not being the case. 

```
// const _fields = unref(fields);
const _orginal_fields = unref(fields);

const _collection = unref(collection);

if (!_orginal_fields || _orginal_fields.length === 0 || !_collection) return aliasedFields;

// Add base keys for any nested fields that are missing
const baseKeys = new Set<string>(_orginal_fields);

_orginal_fields.forEach(field => {
	const fieldParts = field.split('.');

	if (fieldParts.length > 1) {
		const baseKey = fieldParts[0];
		baseKeys.add(baseKey); // Ensure the base key is included
	}
});

// Convert the set back to an array for use in the rest of the function
const _fields = Array.from(baseKeys);
```

Overall, this feels like quite a hacky solution to me and I'm sure it can be much improved by someone more knowledgeable with the intricacies of directus, but it's enough for me and should definitely show that it's possible to implement such a feature, that a lot of people seem to want. For anyone interested I might open a pull request or maybe create a fork with the changes I've made to the code base.

**Reply by hsb-tonmoy** [reply]
> Update on this: I've figured out how to get it working better now and have a nicer interface, it looks similar to the default translation preview now. It also works now when it's in another table through a relation. I've also fixed the issue that you needed to include other fields in the view, more on this below. The only thing that is left for me is to fix the view when displayed in a o2m table view (here it will still show 2 Elements since it seems to be a different rendering mechanism), but I guess this is a rather specific use-case. Here's a small video of what it looks like now and I'm pretty happy with the way it is right now:
> Screen.Recording.2024-10-30.at.01.18.57.mov
> 
> In addition to the changes proposed in #19876, I've also added this code to `useAliasFields` in `use-alias-fields.ts`. This will ensure that the related collection is also included in the query, since there seemed to be issues with this not being the case.
> 
> ```
> // const _fields = unref(fields);
> const _orginal_fields = unref(fields);
> 
> const _collection = unref(collection);
> 
> if (!_orginal_fields || _orginal_fields.length === 0 || !_collection) return aliasedFields;
> 
> // Add base keys for any nested fields that are missing
> const baseKeys = new Set<string>(_orginal_fields);
> 
> _orginal_fields.forEach(field => {
> 	const fieldParts = field.split('.');
> 
> 	if (fieldParts.length > 1) {
> 		const baseKey = fieldParts[0];
> 		baseKeys.add(baseKey); // Ensure the base key is included
> 	}
> });
> 
> // Convert the set back to an array for use in the rest of the function
> const _fields = Array.from(baseKeys);
> ```
> 
> Overall, this feels like quite a hacky solution to me and I'm sure it can be much improved by someone more knowledgeable with the intricacies of directus, but it's enough for me and should definitely show that it's possible to implement such a feature, that a lot of people seem to want. For anyone interested I might open a pull request or maybe create a fork with the changes I've made to the code base.

I'm quite new to Directus, so this might be a pretty basic question, but are you making all these changes directly in the codebase or using it as a plugin? If the former, wouldn't an update revert all the changes?

**Reply by louisprp** [reply]
Yes this is kind of a proposal of changes to be made to the codebase directly. However, I can use it perfectly fine for me, and when official updates come through I could just merge them, as long as they don’t conflict with my changes. This is however not really the purpose of this and it would be much easier if it was merged into the official codebase for everyone to use.

**Comment by DamienMrtl** [comment]
@louisprp Hello, thanks for helping on this issue, I really hope we will be able to make a new successful PR. I created a new branch on my fork (https://github.com/DamienMrtl/directus/tree/local_fields) to add the new changes you added in `use-alias-fields.ts`. However, when running the tests using `pnpm --workspace-root test` I get an error:

```
 FAIL  src/composables/use-alias-fields.test.ts > useAliasFields > aliasing no collisions
│ AssertionError: expected { id: { key: 'id', …(3) }, …(2) } to deeply equal { id: { aliased: false, …(3) }, …(1) }
│ - Expected
│ + Received
│   Object {
│ +   "3468cda4": Object {
│ +     "aliased": true,
│ +     "fieldAlias": "3468cda4",
│ +     "fieldName": "image",
│ +     "fields": Array [
│ +       "3468cda4.url",
│ +     ],
│ +     "key": "image.url",
│ +   },
│ +   "5faa95b": Object {
│ +     "aliased": true,
│ +     "fieldAlias": "5faa95b",
│ +     "fieldName": "image",
│ +     "fields": Array [
│ +       "5faa95b.url.file",
│ +       "5faa95b.alt",
│ +     ],
│ +     "key": "image",
│ +   },
│     "id": Object {
│       "aliased": false,
│       "fieldName": "id",
│       "fields": Array [
│         "id",
│       ],
│       "key": "id",
│ -   },
│ -   "image.url": Object {
│ -     "aliased": false,
│ -     "fieldName": "image",
│ -     "fields": Array [
│ -       "image.url",
│ -     ],
│ -     "key": "image.url",
│     },
│   }
```

I'm not sure if we should adapt something from the test file or your added code.

Please anyone willing to help, feel free to help me on my forked version (https://github.com/DamienMrtl/directus/tree/local_fields) so we can be ready to create a new PR. 

@rijkvanzanten Please can you create an issue for this accepted feature request so we can create a new PR as described here (https://docs.directus.io/contributing/pull-request-process.html#implementing-an-accepted-feature-request).

**Reply by louisprp** [reply]
Alright, I've figured out the cause of the error and the tests for the composables tests run fine now without issues. I'll begin adding my code to your fork.

**Reply by louisprp** [reply]
@DamienMrtl I've created a pull-request for an example inline-translations display, similar to the one I've shown in the video. From what I can tell, it should be usable in different scenarios (for example now also in a table) and I haven't encountered any issues with it yet. As a next step, I'll look into offering more options to configure, but I would believe there isn't that much more that would have to be implemented in that regard. Hopefully something like this can make its way into the repo, at least the option of fetching the other localFields certainly opens up many other possibilities beyond the current scope of this discussion. From what I've tried out, these changes should also not conflict with any commits made to the upstream directus repo, so for anyone interested in using this, they could use the code from the pull-request and build a docker image or run it however needed.

**Comment by leo29plns** [comment]
Hello, any update on this ?

**Reply by wottpal** [reply]
Also waiting for this... By far the most annoying part of Directus for everyone in our team.

**Comment by nickwph** [comment]
Hello! Just checking if there is any updates and will the code be merged anytime soon? My team has been waiting for it. Please feel free to share if there is any concerns. 😄

**Comment by IW-viz** [comment]


**Reply by wottpal** [reply]
Amazing!! 🎉 Is this also planned to be released via the marketplace?

A bit embarrassing though that such a crucial component must be implemented externally.

**Comment by dbargaud** [comment]
Hi,

this is great! Thank you for your contribution.

To be even better, could you add the possibility to have dynamic values in the language code field ?
I want to be able to put something like "{{posts_postId.country.defaultLanguage}}":
![image](https://github.com/user-attachments/assets/f6c2fd28-7642-49a3-9402-35f3344fccbe)

**Reply by dbargaud** [reply]
It does not work if you display more than one translatable field. Is this normal ?

**Comment by IW-viz** [comment]


**Reply by dbargaud** [reply]
Hi,

this is wrong to use this :
```
// Get the relation field name based on collection name
const getRelationFieldName = (collectionName: string): string => {
  const baseName = collectionName.split("_")[0];
  return `${baseName}_id`;
};
```

The collection reference :
```
{{posts.123.language}}
{{articles.456.defaultLanguage}}
{{products.789.locale}}
```

is not a good idea, in directus you do not specify an "id" in this kind of expression.
You must use the api to retrieve the value contained between {{ }}. The expression use a field from the "current record".

if you have 3 collections like this :

- collection "posts_translations"
{
 "posts_translations_id": UUID,
  "posts_postId" : relation to the "posts" collection,
  "languages_code": string,
  "title": string,
  "text": very long text 
}

- collection "posts"
{
  postId:UUID,
  country: relation to the "countries" collection
}

- collection "countries"
{
  "code":string,
  "defaultLanguage":string // containing the language code (en-US or something else)
}

When you specifiy {{posts_postId.country.defaultLanguage}} for the language, you must retrieve by the api using fields:["posts_postId.country.defaultLanguage"], the api will resolve it and find the field "defaultLanguage" by traversing the different relations (posts_translations.posts_postId, posts.country, country.defaultLanguage).

In this example, the url would be /items/posts_translations?fields[]=posts_postId.country.defaultLanguage you just have to add the filter to match the corresponding information to your field value.

I hope my explanation is sufficiently clear :P

**Comment by IW-viz** [comment]


**Reply by dbargaud** [reply]
It looks great ! Exactly what we need.

**Reply by IW-viz** [reply]
hi @dbargaud , I'm looking for a solution to latency. If I have 3 fields with the extension and 200 items, or 600 extensions, I have latency, and it annoys me. I need to find a solution before publishing on npm. Any suggestions?

**Reply by dbargaud** [reply]
How many PM2 instances do you use for your directus ? I guess it comes from the number of api call, 3x200= 600 api calls ?

**Reply by IW-viz** [reply]
I only use a single instance of Directus, without PM2 to manage multiple instances. The large number of API requests you observe actually comes from a different phenomenon:
My collection "pages" has 200 items, and each page displays 3 translated fields, which means there are 600 instances of the language-display extension running on the page. Each of these components makes its own independent API requests.
Currently, each component instance can make:
An initial request to retrieve translations (1 request)
A request to resolve dynamic language codes (1 request)
Preloading requests for common languages (up to 5 requests for the default preloaded languages)
In the worst case, this could result in up to 7 API calls per component. With 600 components, that's potentially 4,200 API calls in total! Even with some caching, the number remains very high.
I'm currently working on an optimization that involves implementing a global cache shared between all components. This should significantly reduce the number of API requests, as components will be able to reuse data already retrieved by other components.
The problem is therefore not related to multiple server instances, but rather to the client architecture of the extension that doesn't efficiently share data between components.

**Reply by dbargaud** [reply]
Ok, that's what I thought too, it will makes a lot of call. Your global cache will solve the session level but there will still be issues with the user concurrency... Meaning if you have 10 users you will have 4,200 * 10 API calls. I am guessing it will be difficult to handle without a customized endpoint to centralize the cache and it will be difficult to handle with the security part. :P

**Comment by IW-viz** [comment]
@louisprp replied yesterday too, https://github.com/directus/directus/discussions/19761#discussioncomment-12466100

**Comment by louisprp** [comment]
To all still interested in this, I have built some very usable components based on the changes proposed by @DamienMrtl and me. The code for the component below can be found in the pull-request. Note however that this requires modification to the source-code of directus, so until the pull-request can be merged into the official repo, the only way to use this is to rely on the fork. If this is fine for you, the display component I've created should work fine. I've tested it for various scenarios and believe I've found a nice compromise, where there shouldn't be any other current features affected or broken, while opening up new possibilities through the addition of "localFields" for other extensions. It accounts for common scenarios and accounts for being displayed in different contexts. (For example when viewing just the ..._translations table, it just returns the string)

Below is a video showcasing the new component. It is based on the translation display and shows translated values in the exact same way. 


https://github.com/user-attachments/assets/d1748d3f-0652-40ff-b90e-7c4005a6824a

**Comment by IW-viz** [comment]
![image](https://github.com/user-attachments/assets/c1341ac7-3869-4095-9ee3-2a899eb854e3)
![image](https://github.com/user-attachments/assets/c59e0728-4a64-4039-9678-a89c2ef196fa)

![image](https://github.com/user-attachments/assets/d1ed5b31-51f6-473a-9210-96ba61faa1af)



# Display of translated fields in Directus v11.5.1

## Problem Context

The GitHub discussion [#19761](https://github.com/directus/directus/discussions/19761) raises an issue regarding the display of translated fields in Directus. According to the initial request, it was not possible to display a translated field in the user's language in its own column, but rather a comma-separated list of all translated values.

## Native Solution in Directus v11.5.1

To understand the problem, everyone has seen this page https://docs.directus.io/guides/headless-cms/content-translations.html, which explains how to create a translation collection. Some of you have found this page https://directus.io/docs/configuration/translations, which is even worse because it mixes the two translation concepts in Directus: word translations and collection management with languages!

There's a missing piece in this documentation!

## Directus v11.5.1 does it naturally!

You shouldn't use the title field in the table but the translation type field and apply a display on it that only exists if it's a translation type field!

So we have an interface that allows displaying a title in the desired language and, as a bonus, there are directly the other translations, all natively in Directus! What a waste of time!

For example, I created a page collection and created a translation field that I called base. Directus automatically created a pages_base collection. In pages_base I add my title and slug fields.

I go back to my pages collection and on the base field of translation type I apply a translations display!

![image](https://github.com/user-attachments/assets/2be07146-837c-441d-aed3-03a34ae86ff7)

**Reply by IW-viz** [reply]
@dbargaud, now we can create a good alternative to use {{posts_postId.country.defaultLanguage}}, need time to do it

**Comment by IW-viz** [comment]
that s work, with no hack ! DIRECTUS so big ! i love, one query(api call) !
![image](https://github.com/user-attachments/assets/b6fd8c8d-db72-4c09-98e1-746e41efa4fe)
![image](https://github.com/user-attachments/assets/3c4c6702-7e45-48d1-92da-4a39f8edcc1d)

**Reply by dbargaud** [reply]
Great ! Where can we find the code ?

**Reply by IW-viz** [reply]
https://github.com/idee-web/directus-extension-display-translations-context

**Reply by IW-viz** [reply]
The extension is now available on npm. Any feedback on compatibility with different Directus versions would be greatly appreciated.

**Reply by dbargaud** [reply]
Ok, I quickly reviewed the extension and unfortunately that is not what I've hoped for.
It seems that you took a different direction.
In the beginning, you were trying to create a display extension which could be use on each translation field. 
What I see now, is finally a variation of what is already existing in directus. 
Surely, we can specify the language to be used to display the value but we are still not able to directly use more than one translation field (meaning one grid column for each wanted field).

Anyway, thank your for your contribution! 😄

**Reply by IW-viz** [reply]
I know! I started at the beginning on an extension that applies to the field that contains the translations like title, but too many API calls, not manageable. I made this one that allows me to have yes a single translation field that returns X fields but in the same column with the processing. I am working on the step after a table layout that keeps a single API call of the collection and to do the processing on the columns, longer to do and other problem, but I do not give up, I need it.
